File: csharp/src/Vista.SDK/GmodNode.cs
using System.Text;
#if NET8_0_OR_GREATER
using System.Collections.Frozen;
#endif

namespace Vista.SDK;

public record class GmodNode
{
    public string Code { get; init; }
    public Location? Location { get; internal init; }

    public VisVersion VisVersion { get; }

    public GmodNodeMetadata Metadata { get; }

    internal readonly List<GmodNode> _children;
#if NET8_0_OR_GREATER
    internal FrozenSet<string> _childrenSet;
#endif
    internal readonly List<GmodNode> _parents;

    public IReadOnlyList<GmodNode> Children => _children;

    public IReadOnlyList<GmodNode> Parents => _parents;

    internal GmodNode(VisVersion version, GmodNodeDto dto)
    {
        VisVersion = version;
        Code = dto.Code;
        Metadata = new(
            dto.Category,
            dto.Type,
            dto.Name,
            dto.CommonName,
            dto.Definition,
            dto.CommonDefinition,
            dto.InstallSubstructure,
            dto.NormalAssignmentNames ?? new Dictionary<string, string>(0)
        );
        _children = new List<GmodNode>();
#if NET8_0_OR_GREATER
        _childrenSet = FrozenSet<string>.Empty;
#endif
        _parents = new List<GmodNode>();
    }

    internal GmodNode WithoutLocation() => Location is null ? this : this with { Location = null };

    internal GmodNode WithLocation(string location)
    {
        var locations = VIS.Instance.GetLocations(VisVersion);

        return this with
        {
            Location = locations.Parse(location)
        };
    }

    internal GmodNode TryWithLocation(string? locationStr)
    {
        var locations = VIS.Instance.GetLocations(VisVersion);
        if (!locations.TryParse(locationStr, out var location))
            return this;

        return WithLocation(location);
    }

    internal GmodNode TryWithLocation(string? locationStr, out ParsingErrors errors)
    {
        var locations = VIS.Instance.GetLocations(VisVersion);

        if (!locations.TryParse(locationStr, out var location, out errors))
            return this;

        return WithLocation(location);
    }

    internal GmodNode WithLocation(in Location location) => this with { Location = location };

    internal GmodNode TryWithLocation(in Location? location)
    {
        if (location is null)
            return this;

        return WithLocation(location.Value);
    }

    internal bool IsIndividualizable(bool isTargetNode = false, bool isInSet = false)
    {
        if (Metadata.Type == "GROUP")
            return false;
        if (Metadata.Type == "SELECTION")
            return false;
        if (IsProductType)
            return false;
        if (Metadata.Category == "ASSET" && Metadata.Type == "TYPE")
            return false;
        if (IsFunctionComposition)
            return Code[Code.Length - 1] == 'i' || isInSet || isTargetNode;
        return true;
    }

    public bool IsFunctionComposition =>
        (Metadata.Category == "ASSET FUNCTION" || Metadata.Category == "PRODUCT FUNCTION")
        && Metadata.Type == "COMPOSITION";

    public bool IsMappable
    {
        get
        {
            if (ProductType is not null)
                return false;
            if (ProductSelection is not null)
                return false;
            if (IsProductSelection)
                return false;
            if (IsAsset)
                return false;

            var lastChar = Code[Code.Length - 1];
            return lastChar != 'a' && lastChar != 's';
        }
    }

    public bool IsProductSelection => Gmod.IsProductSelection(Metadata);

    public bool IsProductType => Gmod.IsProductType(Metadata);

    public bool IsAsset => Gmod.IsAsset(Metadata);

    public GmodNode? ProductType
    {
        get
        {
            if (_children.Count != 1)
                return null;

            if (!Metadata.Category.Contains("FUNCTION"))
                return null;

            var child = _children[0];
            if (child.Metadata.Category != "PRODUCT")
                return null;

            if (child.Metadata.Type != "TYPE")
                return null;

            return child;
        }
    }

    public GmodNode? ProductSelection
    {
        get
        {
            if (_children.Count != 1)
                return null;

            if (!Metadata.Category.Contains("FUNCTION"))
                return null;

            var child = _children[0];
            if (!child.Metadata.Category.Contains("PRODUCT"))
                return null;

            if (child.Metadata.Type != "SELECTION")
                return null;

            return child;
        }
    }

    public bool IsChild(GmodNode node) => IsChild(node.Code);

    public bool IsChild(string code)
    {
#if NET8_0_OR_GREATER
        return _childrenSet.Contains(code);
#else
        for (int i = 0; i < _children.Count; i++)
        {
            if (_children[i].Code == code)
                return true;
        }

        return false;
#endif
    }

    public virtual bool Equals(GmodNode? other) => Code == other?.Code && Location == other?.Location;

    public override int GetHashCode() => HashCode.Combine(Code, Location);

    public sealed override string ToString() => Location is null ? Code : $"{Code}-{Location}";

    public void ToString(StringBuilder builder)
    {
        if (Location is null)
        {
            builder.Append(Code);
        }
        else
        {
            builder.Append(Code);
            builder.Append('-');
            builder.Append(Location);
        }
    }

    internal void AddChild(GmodNode child) => _children.Add(child);

    internal void AddParent(GmodNode parent) => _parents.Add(parent);

    internal void Trim()
    {
        _children.TrimExcess();
        _parents.TrimExcess();
#if NET8_0_OR_GREATER
        _childrenSet = _children.Select(c => c.Code).ToFrozenSet(StringComparer.Ordinal);
#endif
    }

    public bool IsLeafNode => Gmod.IsLeafNode(Metadata);

    public bool IsFunctionNode => Gmod.IsFunctionNode(Metadata);

    public bool IsAssetFunctionNode => Gmod.IsAssetFunctionNode(Metadata);

    public bool IsRoot => Code == "VE";
}

public sealed record class GmodNodeMetadata(
    string Category,
    string Type,
    string Name,
    string? CommonName,
    string? Definition,
    string? CommonDefinition,
    bool? InstallSubstructure,
    IReadOnlyDictionary<string, string> NormalAssignmentNames
)
{
    public string FullType { get; } = $"{Category} {Type}";
}

File: js/src/vista-sdk/lib/GmodNode.ts
import { GmodNodeDto, Gmod } from ".";
import { GmodNodeMetadata } from "./types/GmodNode";
import { naturalSort } from "./util/util";

export class GmodNode {
    private _id: string;
    public code: string;
    public metadata: GmodNodeMetadata;
    private _location: string;
    private _children: GmodNode[];
    private _parents: GmodNode[];

    private constructor(
        id: string,
        code: string,
        metadata: GmodNodeMetadata,
        location?: string,
        parents?: GmodNode[],
        children?: GmodNode[]
    ) {
        this._id = id;
        this.code = code;
        this._location = location || "";
        this.metadata = metadata;

        this._children = children ?? [];
        this._parents = parents ?? [];
    }

    public static createFromDto(dto: GmodNodeDto) {
        const {
            code,
            category,
            type,
            name,
            commonName,
            commonDefinition,
            definition,
            installSubstructure,
            normalAssignmentNames,
            location,
            id = code,
        } = dto;

        return new GmodNode(
            id,
            code,
            {
                category,
                type,
                name,
                commonName: commonName ?? undefined,
                definition: definition ?? undefined,
                commonDefinition: commonDefinition ?? undefined,
                installSubstructure: installSubstructure ?? undefined,
                normalAssignmentNames: new Map<string, string>(
                    Object.entries(normalAssignmentNames ?? {})
                ),
            },
            location
        );
    }

    public withLocation(_location: string) {
        return new GmodNode(
            this._id,
            this.code,
            this.metadata,
            _location,
            this.parents,
            this.children
        );
    }

    public withoutLocation(): GmodNode {
        return new GmodNode(
            this._id,
            this.code,
            this.metadata,
            undefined,
            this.parents,
            this.children
        );
    }

    public get id() {
        return this._id;
    }

    public set location(location: string) {
        this._location = location;
    }

    public get location(): string {
        return this._location;
    }

    public get children() {
        return this._children;
    }

    public get parents() {
        return this._parents;
    }

    public get isMappable() {
        if (!!this.productType) return false;
        if (!!this.productSelection) return false;
        if (this.isProductSelection) return false;
        if (this.isAsset) return false;

        const lastChar = this.code.charAt(this.code.length - 1);
        return lastChar !== "a" && lastChar !== "s";
    }

    public get isProductSelection() {
        return Gmod.isProductSelection(this.metadata);
    }

    public get isProductType() {
        return Gmod.isProductType(this.metadata);
    }

    public get isAsset() {
        return Gmod.isAsset(this.metadata);
    }

    public get productType(): GmodNode | undefined {
        if (this._children.length !== 1) return;
        if (!this.metadata.category.includes("FUNCTION")) return;

        const child = this._children[0];
        if (child.metadata.category !== "PRODUCT") return;
        if (child.metadata.type !== "TYPE") return;
        return child;
    }

    public get productSelection(): GmodNode | undefined {
        if (this._children.length != 1) return;
        if (!this.metadata.category.includes("FUNCTION")) return;

        const child = this._children[0];
        if (!child.metadata.category.includes("PRODUCT")) return;
        if (child.metadata.type !== "SELECTION") return;

        return child;
    }

    public isChild(node: GmodNode | string): boolean {
        return typeof node === "string"
            ? this.isChildFromCode(node)
            : this.isChildFromCode(node.code);
    }

    private isChildFromCode(code: string) {
        for (let i = 0; i < this._children.length; i++) {
            if (this._children[i].code == code) return true;
        }

        return false;
    }

    public equals(other?: GmodNode): boolean {
        if (!other) return false;
        return this.code === other.code && this._location === other._location;
    }

    public toString(): string {
        return this._location ? `${this.code}-${this._location}` : this.code;
    }

    public addChild(child: GmodNode) {
        this._children.push(child);
        this._children = this._children.sort((a, b) =>
            naturalSort(this.getSortedString(a), this.getSortedString(b))
        );
    }

    public addParent(parent: GmodNode) {
        this._parents.push(parent);
        this._parents = this._parents.sort((a, b) =>
            naturalSort(this.getSortedString(a), this.getSortedString(b))
        );
    }

    private getSortedString(node: GmodNode) {
        return node.location ? node.code + "-" + node.location : node.code;
    }

    public get isLeafNode(): boolean {
        return Gmod.isLeafNode(this.metadata);
    }
    public get isFunctionNode(): boolean {
        return Gmod.isFunctionNode(this.metadata);
    }
    public clone(): GmodNode {
        return new GmodNode(
            this._id,
            this.code,
            { ...this.metadata },
            this._location,
            [...this._parents],
            [...this._children]
        );
    }
}
